### ISSUE TITLE
Provide guidance on what sort of tests should be written

### ISSUE THREAD

**Issue by wagenet** [typings/generator-typings#61]
I'd love to write some tests for my typings but I'm sort of at a loss as to what and how I should be testing.

**Comment by unional** [comment]
There are two ways you can test your typings:
- Reusing tests from the source
- Writing tests yourself.

Reusing tests from the source is the best approach because all tests are written for you. You only need to make minor changes to them.
A drawback is that it might not be easy to do because the source can have complicated test configuration.

One great example for this is https://github.com/typed-typings/npm-chai

I'll do a write up in the future on this topic.

On writing tests yourself, the generator installs `blue-tape` for you so you can write test using it.
If you can test functionality, it would be great. This would be the same as writing test against the source.
If the environment is complicated and have dependencies hard to decoupled (e.g. the source is a UI library and test requires mocking user interaction), then testing against the shape should be fine too.

It really depends on how far you want to test.
Examples:
https://github.com/typed-typings/npm-onchange
(I'll update that repo to match the latest structure).

**Comment by wagenet** [comment]
Thanks for the reply. What's the point of writing these tests? To make sure that you're building against a valid source?

**Comment by unional** [comment]
Make sure the typings are written correctly, of course. :smile:
Since there is no type in javascript, even if you create a wrong signature, you won't detect it until runtime.

e.g.
```js
// source code
function foo(something) {
  return something++;
}
```
The source code expects `something` to be a number. But in your typings:
```ts
function foo(something: string): string;
```
It won't fail until you use it. i.e.
```ts
// consumer program
import foo ....somehow

let x = "abc";
foo(x);
```

Because your typings provide guidance to the consumer, they will write the code that way and will fail when they run it.
`tsc` will compile fine.

By the way, instead of https://github.com/typed-typings/npm-onchange
Look at https://github.com/typed-typings/npm-bluebird
`npm-onchange` will be removed as the typings is added to the source already.

**Comment by wagenet** [comment]
I feel like I'm probably being pretty dense here but I'm still not 100% sure I follow. Is the idea to manually write your own tests as if you were a new user following the guidance from the typings file just to make sure things actually run when you follow the guidance of the typings?

**Comment by unional** [comment]
Yes, you are correct.

Essentially you are writing unit test for the library.
That's why I would like to use the tests from source if available.

In the future, if https://github.com/Microsoft/TypeScript/issues/7661 is available, when we can run the tests directly in the source through `tsc` and we don't need to write any tests. :rose:

**Comment by unional** [comment]
Of course, you don't really need to write all unit tests for the library. Just those that can cover the type checking is sufficient.

**Comment by unional** [comment]
I notice you are working on typed-ember. I might able to help to start a few tests for you.

**Comment by unional** [comment]
Another example for re-using tests in source:
https://github.com/typed-typings/npm-batch

**Comment by Demurgos** [comment]
> Writing runtime times are needed as JavaScript is not typed to begin with.

> You only need to write runtime tests that ensure the signature is correct.
Checkout generator-typings, there is an issue related to this topic.

> The best is to reuse tests from the source, but that still needs some work. 

(https://github.com/typed-typings/npm-lodash/issues/8)

I understand why typings are important, but `typings` is about the compilation checks, not the runtime. unit-tests. I do not say that functionality tests are not useful, but they should belong to the lib repo.

Wouldn't compiling `let result: number = _.add(1, 2);` without any warning from the compiler be more relevant than running `t.equal(_.add(1, 2), 3);` ? (When it comes to __type__ checks)
I mean, shouldn't tests for types be just a lot of sample code with various type annotations instead of calls to some assertion library ?

**Comment by Demurgos** [comment]
> As in the example above, __it will compile fine__ but it will fail at runtime. 

Exactly, because it was correctly typed. If it was typed as "returning a string" or "accepting an array", it would have emitted a warning.

To test your example above for example, I would write the following test:
````typescript
/*
   Documentation: foo (something)
       Increments the integer something and returns the result
*/
let x: number = 10;
let result: number = foo(x); // test that the return type for a number is a number
````
According to the documentation, `foo` expects a number and returns a number. This is what is stated in the documentation, therefore this what we want to test: our typings should let this code compile without any issue.

This way we can detect:
````typescript
// wrong-typing.ts
function foo(something: string): string;
````
This causes warnings -> there is a problem !

````typescript
// correct-typing.ts
function foo(something: number): number;
````
This does not cause warnings -> maybe there is no problem


I feel like testing type-definitions should require this sort of framework: instead of checking at runtime, it should propose ways to check things like "this code should not emit type warnings during compilation" or "this code should emit type errors during compilation".

For example, I would like to write something like this:
````typescript
import {lib} from "awesome-lib-to-test-typings";

describe("foo typings", function() {

  let fooContext = lib.testDefinition("foo.ts"); // or more likely an array of ts sources, or a config object
  /*
    fooContext has two methods: assertWarning(codeSample: string) and assertOk(codeSample: string)
    They both try to compile the code sample with the context of foo.d.ts.
    The first one throws an error if the compilation is successful, the second one throws an error if the compilation
    emits a warning.
   */
  
  it("should allow normal usage", function() {
    // number argument, number return type
    fooContext.assertOk(`
      let result: string = foo(10);
    `);
    
    // add more tests if the function has multiple signatures
  });

  it("should prevent illegal usage", function() {
    // incorrect arguments
    fooContext.assertWarning(`
      let result: number = foo("string");
    `);
    
    // incorrect return type
    fooContext.assertWarning(`
      let result: string = foo(10);
    `);
  });
});
````

**Comment by blakeembrey** [comment]
@Demurgos I think you're getting a bit deep into it, so let's just review something basic. No, we shouldn't be doing complete runtime tests. However, we should test the runtime. This is simply because of things like, are our `import x = require('...')` correct, etc. Currently, we don't _just_ test the runtime - we're using `ts-node` which is also checking the compilation is 100% correct. Sure, we could just run the compiler over it instead - but isn't it worth running (with the same amount of effort) the actual script we wrote and make sure it's correct (E.g. confirm it's correct, not just correct according to our reading of the docs - which, of course, the docs may not be 100% complete at times)?

**Comment by Demurgos** [comment]
Ok, thank you for reminding me that the tests are using `ts-node` so they are both code samples for compilation warnings and, as a bonus, runtime checks. So yes, now I see how these tests fit into this.

But this is not completely satisfying: `typings` should both allow the correct syntax and prohibit the incorrect syntax. The current tests only do "positive" checks: the value is valid, the type is valid and everything is okay.
Testing the wrong case is also important. For the runtime, the numerous testing framework allow to check for inequalities or throwed errors. What is the equivalent for the types ?
I am trying to come up with a good example

For example, I want to be sure that the user cannot write this (unless he forces it with `any`...):
````typescript
let result: string = foo(10);
````
Is there a way/is it good to test it ?

Another concern I have is if we do some refactoring and `function foo(something: number): number;` ends up as `function foo(something: number): any`. Is it possible to catch it only with tests ? (it seems trickier because of the nature of `any`)

Maybe you're right, I'm going too far by not only wanting to provide interfaces, function signatures, etc. but also wanting to be sure that these typings are strict enough (anyone can add declarations with `any` everywhere).

**Comment by unional** [comment]
> ...and prohibit the incorrect syntax

That is exactly the point for running the tests instead of just compiling them. As said, the example I gave above will compile fine. But it will not run fine because either:
1. We read the doc incorrectly
2. We made a mistake
3. The document is wrong

Until we actually run the test, we will not know.

> For the runtime, the numerous testing framework allow to check for inequalities or throw errors.

Since you are dealing with TypeScript, if the typings is incorrect, you won't able to compile, that is the "negative" test, and it is handled by the compiler.

About `any`, there is no easy way, because you explicitly declare it as `any`, that is a conscious choice.
